### PR TITLE
feat: implement @ mentions with notifications (#517)

### LINKED ISSUE #517: Implement @ mentions with notifications
## Feature Request
Implement @ mention functionality that allows users to mention other users in messages with automatic notifications.

## Current Behavior
- Users can type @ in messages but it doesn't trigger any special functionality
- No mention parsing or detection exists
- No notification system for mentioned users

## Desired Behavior
1. When a user types @ in a message, show a mention autocomplete dropdown with available users
2. Parse @ mentions in message content (e.g., @username)
3. Highlight/style @ mentions in rendered messages
4. Send notifications to mentioned users
5. Mentioned users should see a notification or badge indicating they were mentioned
6. Clicking on a mention should open that user's profile

## Implementation Requirements
- **Frontend:**
  - Message input autocomplete for @ mentions (show matching users)
  - Parse and render @ mentions in messages with special styling/highlighting
  - Link @ mentions to user profiles

- **Backend:**
  - Parse message content for @ mentions during save
  - Store mention metadata (either in message table or separate mentions table)
  - Trigger notifications for mentioned users
  - API to retrieve mentions for a user (notifications)

- **Database:**
  - Consider adding a `mentions` or `messageMentions` table to track user mentions
  - Or store mentions as JSON in Message content if simple approach preferred
  - Add notifications table/system if not already implemented

## Acceptance Criteria
- [ ] Users can type @ to see autocomplete of available users
- [ ] @ mentions are parsed and stored when message is sent
- [ ] Mentioned users receive a notification
- [ ] @ mentions are highlighted/styled differently in rendered messages
- [ ] Clicking on a mention opens user profile or shows user info
- [ ] Mentions work in both regular messages and replies
- [ ] Mention parsing handles edge cases (multiple mentions, mentions in edits, etc.)

### PR DESCRIPTION
**Pull request by FardeenI**
## Summary

Closes #517

Implements the full @ mention flow — autocomplete in the message input, styled rendering in messages, backend mention parsing and storage, and a real-time notification bell.

### Backend
- **New Prisma models**: `MessageMention` + `Notification` with migration
- **`mention.service.ts`**: parses `@username` tokens, resolves to server members, creates mention records and per-user notifications (fire-and-forget, never blocks message delivery)
- **`notification.service.ts`** + **`notification.router.ts`**: tRPC procedures — `getNotifications`, `getUnreadCount`, `markAsRead`, `markAllAsRead`
- **`serverMember.searchMembers`**: prefix-search for autocomplete (caller must be a member)
- **Integration**: `processMentions` called in `sendMessage`, `editMessage`, `createReply`; idempotent via `skipDuplicates`
- **`USER_MENTIONED` event** + **`/api/events/user` SSE endpoint** for real-time notification delivery

### Frontend
- **`MentionAutocomplete.tsx`**: dropdown with full keyboard navigation (↑/↓/Enter/Tab/Escape)
- **`MentionText.tsx`**: renders `@username` tokens as inline pills; self-mentions get accent color
- **`NotificationBell.tsx`**: bell icon with live unread badge, notification panel, mark-as-read, real-time SSE updates
- **`MessageInput.tsx`**: detects `@`, debounces 120 ms, calls `searchMembers`, inserts on selection
- **`MessageItem.tsx`**: renders all message content through `MentionTextWithSelf`
- **`TopBar.tsx`** / **`HarmonyShell.tsx`**: wires `NotificationBell` with authenticated user ID

## Test plan

- [ ] Type `@` in a channel — dropdown appears; typing filters candidates
- [ ] Keyboard navigation: ↑/↓ moves selection, Enter/Tab inserts, Escape closes
- [ ] Sent message shows `@username` as styled pill in the message list
- [ ] Self-mentions use accent background; others use subdued style
- [ ] Mentioned user sees real-time unread count on bell icon (SSE)
- [ ] Bell panel shows recent mentions; mark-as-read and mark-all-read work
- [ ] Mentions work in thread replies
- [ ] Non-member mention silently skips notification
- [ ] Self-mention does not create a notification
- [ ] `npx tsc --noEmit` clean in both backend and frontend

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: harmony-backend/src/services/mention.service.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    const now = new Date();
+
+    // Upsert mentions (idempotent on re-save/edit)
+    await prisma.messageMention.createMany({
+      data: members.map((m) => ({ messageId, userId: m.userId })),
+      skipDuplicates: true,
+    });
+
+    // Create one notification per newly mentioned user
+    const notifications = await prisma.$transaction(
```
Comment (acabrera04) [inline harmony-backend/src/services/mention.service.ts]: `messageMention.createMany(..., skipDuplicates: true)` dedupes mention rows, but notifications are still created unconditionally for every resolved member on every edit/re-save. That contradicts the stated idempotency and will generate duplicate mention notifications for the same message/user pair. Please gate notification creation to newly-inserted mentions only (or enforce a unique constraint and upsert strategy for `(message_id, user_id, type)`).

File: harmony-frontend/src/components/channel/MessageInput.tsx [hunk 2]
```diff
@@ -214,21 +228,108 @@ export function MessageInput({
     onMessageSent,
     pendingAttachments,
     replyingTo,
+    closeMentionDropdown,
```
Comment (acabrera04) [inline harmony-frontend/src/components/channel/MessageInput.tsx:275]: `handleSend` references `closeMentionDropdown` in its dependency list before `closeMentionDropdown` is declared. With `const` this is a temporal-dead-zone/use-before-declaration problem and can break compile/runtime depending on tooling. Move `closeMentionDropdown` above `handleSend` (or refactor so dependencies are declared after initialization).

**Review by acabrera04 — CHANGES_REQUESTED** [review]
### 8-Point Review Checklist
- [ ] 1. Prioritize Logic over Style: `processMentions` is not idempotent for notifications, so repeated edits/replies can spam duplicate mention notifications for the same message/user pair.
- [x] 2. Security First: Mention lookup is server-scoped and authenticated; notification mutators are user-scoped.
- [x] 3. Architectural Alignment: Changes follow existing service/router layering and event bus patterns.
- [x] 4. Issue Completion: Core mention + notification flow is implemented across backend/frontend paths.
- [x] 5. No Nitpicking: Feedback below is only on behavioral/correctness blockers.
- [x] 6. Avoid Repetition: No prior human inline review feedback exists on this PR, so comments below are net-new.
- [x] 7. Iterative Reviews: Reviewed latest head commit `98c09259009c18cda48b1ea2c306205bac9bc415`; no prior unresolved reviewer threads to re-evaluate.
- [ ] 8. Prevent CI Failures: Current PR checks include failing frontend build/tests and integration jobs, and one frontend dependency-order issue here can cause compile/runtime failure.

Bundled review comments are inline below.

**Comment by FardeenI** [comment]
All review comments addressed in de4f067:

**Notification idempotency (Allen's main concern)**
- Added `@@unique([userId, type, messageId])` constraint on `Notification` in `schema.prisma` and `CREATE UNIQUE INDEX "idx_notifications_unique"` in the migration SQL.
- `processMentions` now uses `notification.createMany({ skipDuplicates: true })` — duplicate calls (e.g. on message edit) skip already-notified users at the DB level.
- SSE events are only fired for *newly inserted* notifications (fetched by `createdAt: now` immediately after createMany), not for skipped duplicates.

**Module-level regex (`mention.service.ts`)**
- Replaced the module-level `const MENTION_RE = /.../g` with `content.matchAll(/@([\w]{1,32})/g)` inside the function body. A fresh iterator is created per call so `lastIndex` never bleeds between invocations.

**`MentionTextWithSelf` Rules of Hooks violation**
- Removed `MentionTextWithSelf` entirely. `MentionText` now accepts a `currentUsername?: string` prop. `MessageItem` already had `user` from `useAuth()` and passes `currentUsername={user?.username}` directly.

**Debounce timer cleanup (`MessageInput.tsx`)**
- Added `clearTimeout(mentionDebounceRef.current)` in both the channel-switch `useEffect` and a dedicated unmount cleanup `useEffect`.

**Jest tests (`tests/mention.service.test.ts`)**
- Added tests for `extractMentionedUsernames`: single mention, multiple/dedup, lowercase normalisation, no-bleed across repeated calls, and email-style edge case.
- Added tests for `processMentions`: creates mention + notification rows for valid members, skips self-mentions, skips non-members, is idempotent (two calls → one notification), no-op for mention-free content.

**Review by acabrera04 — APPROVED** [review]
### 8-Point Review Checklist
- [x] 1. Prioritize Logic over Style: New head fixes the prior logic blockers (notification idempotency and mention processing correctness) with DB constraints + deduplicated writes.
- [x] 2. Security First: Auth/member boundaries remain intact; no new secret exposure or authorization regression observed in the incremental changes.
- [x] 3. Architectural Alignment: Changes stay aligned with existing service/router/event patterns and Prisma model conventions.
- [x] 4. Issue Completion: Mention autocomplete/rendering + backend mention persistence + notifications + tests now form a complete implementation for #517.
- [x] 5. No Nitpicking: Decision based on correctness/behavior only; no style-only feedback.
- [x] 6. Avoid Repetition: No duplicate feedback posted; prior comments were re-checked and are now addressed in this head.
- [x] 7. Iterative Reviews: Re-reviewed against latest head `de4f067e0a2fb1543ce9a874852084ae4a7787dd` from previous reviewed head `98c09259009c18cda48b1ea2c306205bac9bc415`; previous blockers are resolved (TDZ/use-before-declare fix, notification uniqueness/idempotency, debounce cleanup, mention parsing cleanup, tests added).
- [x] 8. Prevent CI Failures: Current PR checks are passing (backend/frontend lint+build, backend/frontend tests, integration tests, Vercel).

Final decision: APPROVE.
